src/main.rs

```diff
@@ -716,363 +748,400 @@ fn main() -> Result<(), Box<dyn Error>> {
-                    .with_unit(units.length.name())?
-                    .with_desc("wavelength of the carrier")?
-                    .write(&wavelength.convert(&units.length))?
-                .new_dataset("polarization")?
-                    .with_desc("linear/circular")?
-                    .write(&pol)?
-                .new_dataset("focusing")?
-                    .with_desc("true/false => pulse is modelled in 3d/1d")?
-                    .write(&focusing)?
-                .new_dataset("envelope")?
-                    .with_desc("pulse envelope of the laser vector potential")?
-                    .write(&envelope)?
-                .new_dataset("chirp_b")?
-                    .with_unit("1")?
-                    .with_desc("parameter that appears in carrier phase = phi + b phi^2")?
-                    .write(&chirp_b)?
-                .new_dataset("waist")?
-                    .with_unit(units.length.name())?
-                    .with_desc("radius in the focal plane at which intensity is 1/e^2 of its peak value")?
-                    .with_condition(|| focusing)
-                    .write(&waist.convert(&units.length))?
-                .new_dataset("fwhm_duration")?
-                    .with_unit("s")?
-                    .with_desc("full width at half maximum of the temporal intensity profile")?
-                    .with_condition(|| envelope == Envelope::Gaussian)
-                    .write(&(n_cycles * wavelength / SPEED_OF_LIGHT))?
-                .new_dataset("n_cycles")?
-                    .with_unit("1")?
-                    .with_desc("number of wavelengths corresponding to the total pulse duration")?
-                    .with_condition(|| matches!(envelope, Envelope::CosSquared | Envelope::Flattop))
-                    .write(&n_cycles)?;
-
-            let charge = match species {
-                Species::Electron => (npart as f64) * weight * ELECTRON_CHARGE,
-                Species::Positron => (npart as f64) * weight * -ELECTRON_CHARGE,
-                Species::Photon => 0.0,
-            };
+        for stat in pstats.iter_mut() {
+            stat.evaluate(&world, &positrons, "positron");
+        }
 
-            let r_max = if normally_distributed {
-                max_radius.unwrap_or(std::f64::INFINITY)
-            } else { // uniformly distributed
-                radius
-            };
+        if id == 0 {
+            if !estats.is_empty() || !gstats.is_empty() || !pstats.is_empty() {
+                use std::fs::File;
+                use std::io::Write;
+                let ss: &str = &format!("{:.1}_", a0);
+                let filename = format!("{}{}{}{}{}stats.txt", output_dir, if output_dir.is_empty() {""} else {"/"}, 
+                                                                      ident, if ident.is_empty() {""} else {"_"}, 
+                                                                      if a0_values.len() > 1 {ss} else {""});
```

Comment:
Nice way to ensure that the stats output get different names! To make sure that the same applies to the distribution output, I'd suggest

```suggestion
                                                                      current_ident, if current_ident.is_empty() {""} else {"_"});
```

(i.e. replacing ident with current_ident everywhere) and then adding

```
        let current_ident = if a0_values.len() > 1 {
            format!("{}_{:.3}", ident, a0)
        } else {
            ident.to_owned()
        };
```

above this.